### PR TITLE
#166272099 user signup

### PR DESCRIPTION
**Pull request by mosinmiloluwa01**
#### What does this PR do?
This PR creates signup feature

#### Description of Task to be completed?
- install jswebtoken
- create user controller
- setup folder structure
- write unit test for signup
- write validations for signup
- create aliases for import path
- generate token
- save token to cookie
- create helper functions for cookie, token, trim and lowercase
- create middleware to ensure an email or username does not exist already
- integrate signup controller with signup model
- refactor Ezekiel's test to add slug to title

#### How should this be manually tested?
- run command "npm run dev"
- input "http://localhost:3000/api/v1/auth/signup" post route in postman
- fill in first name, last name, email, password, username
- click on send

#### Any background context you want to provide?
- user must have node, babel-plugin-module-resolver, eslint-import-resolver-babel-module installed
- this feature requires 3 environmental variables: SECRET, TOKENEXPIRYDATE, and COOKIEEXPIRYDATE

#### What are the relevant pivotal tracker stories?
[166272099](https://www.pivotaltracker.com/story/show/166272099)

### CODE REVIEW

File: server/controllers/helperFunctions/tokenHelper.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  /**
+   * Generates new token
+   * @param {Object} id - id
+   * @returns {Object} token - returns token
+   */
+  static tokenGenerator(id) {
+    const payload = {
+      id
+    };
+    const token = jwt.sign(payload, process.env.SECRET, { expiresIn: '24h' });
```
Comment (Eazybee) [inline server/controllers/helperFunctions/tokenHelper.js]: You might want to hide `24h`, we don't want to give hackers clues or make their life easier

File: server/controllers/helperFunctions/tokenHelper.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    return `${token}`;
```
Comment (Eazybee) [inline server/controllers/helperFunctions/tokenHelper.js]: why do you ```return  `${token}`; ``` when you can just `return token;`?

File: server/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    res.cookie(
+      'access_token',
+      { token: tokenHelper.tokenGenerator(newUser.id) },
+      {
+        maxAge: 3600,
+        httpOnly: true,
+        secure: true
+      }
+    );
+    data.push(newUser);
```
Comment (Eazybee) [inline server/controllers/userController.js]: I thought you were supposed to be saving new user information to the database.
All I see is you pushing it back to dummy data.

File: server/dummyData/data.json [hunk 4]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (Eazybee) [inline server/dummyData/data.json]: new line needed here

File: server/tests/index.spec.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import chai from 'chai';
```
Comment (Eazybee) [inline server/tests/index.spec.js]: I don't think you need this import here

File: server/tests/index.spec.js [hunk 6]
```diff
@@ -1,5 +1,6 @@
...
 import chaiHttp from 'chai-http';
```
Comment (Eazybee) [inline server/tests/index.spec.js]: same here

File: server/tests/index.spec.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
...
+import './controller/user.spec';
 
 chai.use(chaiHttp);
```
Comment (Eazybee) [inline server/tests/index.spec.js]: same here

File: server/tests/controller/user.spec.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../../server';
+
+chai.use(chaiHttp);
+
+const { expect } = chai;
+// const token;
```
Comment (Eazybee) [inline server/tests/controller/user.spec.js]: no commented code

File: server/controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    const email = data.find(user => user.email === req.body.email);
+    if (email) {
+      return res.status(409).send({
+        status: 409,
+        message: 'email already exists'
+      });
+    }
+    const newUser = {
+      id: uuid.v4(),
+      email: req.body.email.toLowerCase().trim(),
```
Comment (Eazybee) [inline server/controllers/userController.js]: I think you can write a helper function for this ```toLowerCase().trim()```

something like this 
```
toLowerCaseAndTrim(obj){
   ....
   return newObj;
}
```
You may check here for an efficient way to loop through an object in js https://zellwk.com/blog/looping-through-js-objects/

**Review by Eazybee — CHANGES_REQUESTED** [review]
Good job @mosinmiloluwa01.
Kindly spare some moment to reflect on the reviews

File: server/controllers/userController.js (hunk 3 above)
Comment (mosinmiloluwa01) [reply server/controllers/userController.js]: Hi Ezekiel,

Thank you for the observation.

The dummy data was created pending the time the user model is merged to develop. As instructed by Benny.

File: server/routes/v1/index.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import express from 'express';
+import user from './user';
+
+const v1Router = express.Router();
```
Comment (frankly034) [inline server/routes/v1/index.js]: No need for the v1 in the router variable, its already in the v1 folder.

File: server/controllers/userController.js (hunk 9 above)
Comment (benfluleck) [reply server/controllers/userController.js]: how many lines of code will this save lol

File: server/controllers/helperFunctions/tokenHelper.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import jwt from 'jsonwebtoken';
```
Comment (frankly034) [inline server/controllers/helperFunctions/tokenHelper.js]: I think this file should be in a helpers folder in the root and should be called token.js.
```server/helpers/token.js```

File: server/controllers/userController.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import uuid from 'uuid';
+import data from '../dummyData/data.json';
```
Comment (frankly034) [inline server/controllers/userController.js]: Change the name from data to newUser and it should be coming from ```server/fixtures/user.js```

File: server/controllers/userController.js (hunk 9 above)
Comment (Eazybee) [reply server/controllers/userController.js]: I was thinking of  Single responsibility.
The S in Solid  :-)

File: package.json [hunk 13]
```diff
@@ -36,7 +37,9 @@
     "request": "^2.87.0",
     "sequelize": "^5.8.6",
     "sequelize-cli": "^5.4.0",
-    "underscore": "^1.9.1"
+    "underscore": "^1.9.1",
```
Comment (benfluleck) [inline package.json]: Remove this library

File: server/controllers/userController.js (hunk 9 above)
Comment (benfluleck) [reply server/controllers/userController.js]: Ok i misunderstood @Eazybee  what you wrote from your helper function. Please be more explicit when giving an example so nothing gets lost.

@mosinmiloluwa01  since you have access to the `req.body` you can find a way to loop through the values of this object and then lowercase and trim each value. However, remember not all fields need to be lowerCased and trimmed.

File: server/routes/index.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
+import v1Router from './v1';
```
Comment (benfluleck) [inline server/routes/index.js]: what is v1Router or folder v1.

File: server/routes/v1/index.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+v1Router.use('/auth', user);
```
Comment (benfluleck) [inline server/routes/v1/index.js]: `v1Router` should really be `authRouter`

File: server/routes/v1/index.js [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import express from 'express';
+import user from './user';
```
Comment (benfluleck) [inline server/routes/v1/index.js]: maybe auth

File: server/routes/v1/user.js [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
+import userController from '../../controllers/userController';
```
Comment (benfluleck) [inline server/routes/v1/user.js]: authController

File: server/tests/controller/user.spec.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+describe('User signup', () => {
+  it('should create a new user and save a token in the cookie', (done) => {
+    chai
+      .request(app)
+      .post('/api/v1/auth/signup')
+      .send({
+        email: 'testing@authorshaven.com',
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js]: Put this in a fixtures folder

**Review by Eazybee — APPROVED** [review]

**Review by akinyeleolat — APPROVED** [review]

File: server/middlewares/userValidations.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+  const {
+    username, firstName, lastName, email, password
+  } = request.body;
+
+  const data = {
+    firstName: firstName && typeof firstName === 'string' ? firstName.trim() : firstName,
+    lastName: lastName && typeof lastName === 'string' ? lastName.trim() : lastName,
+    email,
+    username: username && typeof username === 'string' ? username.trim() : username,
+    password: password && password.trim()
```
Comment (frankly034) [inline server/middlewares/userValidations.js]: I think passwords can have trailing white spaces and might not require trimming.

**Review by frankly034 — COMMENTED** [review]
Nice work using dotenv, an env.example file would come in handy, so we can know out of the box the variables we need to define.

File: server/controllers/userController.js (hunk 12 above)
Comment (mosinmiloluwa01) [reply server/controllers/userController.js]: This has been fixed. that was my former PR

File: server/controllers/helperFunctions/tokenHelper.js (hunk 11 above)
Comment (mosinmiloluwa01) [reply server/controllers/helperFunctions/tokenHelper.js]: This has been fixed. that was my former PR

File: server/routes/v1/index.js (hunk 10 above)
Comment (mosinmiloluwa01) [reply server/routes/v1/index.js]: This has been fixed. that was my former PR

File: .gitignore [hunk 20]
```diff
@@ -44,4 +44,7 @@ node_modules
 *.sqlite
 
 #nyc_output
-.nyc_output
+.nyc_output/
+
+#package-lock.json
+package-lock.json
```
Comment (benfluleck) [inline .gitignore]: remove this it is needed

File: server/fixtures/data.js [hunk 21]
```diff
@@ -0,0 +1 @@
+export default [];
```
Comment (benfluleck) [inline server/fixtures/data.js]: What is this file doing here?. It is not exporting anything and it is named to generically.

File: server/helpers/utils.js [hunk 22]
```diff
@@ -1,17 +1,45 @@
...
  * @return {string} unique_slug
  */
-const createUniqueSlug = title => `${slug(title, { lower: true })}-${Date.now()}`;
 
 /**
  * @param {string} password
  * @return {string} hash
  */
+const { SECRET, TOKENEXPIRYDATE, COOKIEEXPIRYDATE } = process.env;
+export const hashPassword = password => bcrypt.hashSync(password, bcrypt.genSaltSync(10));
```
Comment (benfluleck) [inline server/helpers/utils.js:9]: space out your function and definition. It helps readability

File: server/middlewares/existingUserChecker.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+import { toLowerCaseAndTrim } from '<helpers>/utils';
+
+const { User } = model;
+const userCheck = async (req, res, next) => {
+  const { email, username } = req.body;
+  const formattedValues = toLowerCaseAndTrim(req.body);
+  const userEmailExist = await User.findOne({ where: { email } });
+  const userNameExist = await User.findOne({ where: { username } });
+  let message;
+  if (userEmailExist && userNameExist) {
```
Comment (benfluleck) [inline server/middlewares/existingUserChecker.js:17]: space out the function logic and your variable definitions

File: server/middlewares/existingUserChecker.js [hunk 24]
```diff
@@ -0,0 +1,27 @@
+import model from '../models';
+import { toLowerCaseAndTrim } from '<helpers>/utils';
+
+const { User } = model;
+const userCheck = async (req, res, next) => {
```
Comment (benfluleck) [inline server/middlewares/existingUserChecker.js]: change to a better name like `validUserCheck`

File: server/models/user.js [hunk 25]
```diff
@@ -1,6 +1,5 @@
-import utils from '../helpers/utils';
+import { hashPassword } from '../helpers/utils';
```
Comment (benfluleck) [inline server/models/user.js]: Please resolve this with the new syntax `<helpers>`

File: server/tests/models/user.spec.js [hunk 26]
```diff
@@ -24,7 +23,7 @@ const modelTest = describe('Model - User', () => {
         expect(user.firstName).to.equal(newUser.firstName);
         expect(user.bio).to.equal(newUser.bio);
       } catch (error) {
-        console.log(error);
+        console.log(newUser);
```
Comment (benfluleck) [inline server/tests/models/user.spec.js]: Why are you console.logging user ?

File: server/tests/models/user.spec.js [hunk 27]
```diff
@@ -92,7 +91,7 @@ const modelTest = describe('Model - User', () => {
         expect(user.password).to.not.equal(undefined);
         expect(user.password).to.not.equal(newUser1.password);
       } catch (error) {
-        console.log(error);
+        return error;
```
Comment (benfluleck) [inline server/tests/models/user.spec.js]: Why are you returning error are you re-using it somewhere else

File: server/tests/models/user.spec.js [hunk 28]
```diff
@@ -81,7 +80,7 @@ const modelTest = describe('Model - User', () => {
         const user = await User.create(newUser1);
         expect(user.email).to.equal(newUser1.email);
       } catch (error) {
-        console.log(error);
+        return error;
```
Comment (benfluleck) [inline server/tests/models/user.spec.js]: Why are you returning error are you re-using it somewhere else

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/tests/models/article.spec.js [hunk 29]
```diff
@@ -12,7 +12,7 @@ before(async () => {
   try {
     await sequelize.sync({ force: true });
   } catch (error) {
-    console.log(error);
+    return error;
```
Comment (benfluleck) [inline server/tests/models/article.spec.js]: Why are you returning error here?

File: server/fixtures/user.js [hunk 30]
```diff
@@ -1,9 +1,26 @@
...
+  password: 'testing'
+};
+export const emptyUserData = {
+  email: '',
+  username: '',
+  lastName: '',
+  firstName: '',
+  password: ''
+};
+export const badUserData = args => ({
```
Comment (benfluleck) [inline server/fixtures/user.js]: This is a function, I expect it to be named like on. Maybe `getUSerData`

Comment (mosinmiloluwa01) [reply server/fixtures/user.js]: Done.

File: server/helpers/utils.js [hunk 31]
```diff
@@ -1,17 +1,47 @@
 import slug from 'slug';
 import bcrypt from 'bcryptjs';
+import jwt from 'jsonwebtoken';
+import env from 'dotenv';
```
Comment (benfluleck) [inline server/helpers/utils.js]: why is this here you should only be required to pass this is your entry file

File: server/fixtures/user.js [hunk 32]
```diff
@@ -1,9 +1,26 @@
...
+  firstName: 'testing',
+  password: 'testing'
+};
+export const emptyUserData = {
+  email: '',
+  username: '',
+  lastName: '',
+  firstName: '',
+  password: ''
+};
```
Comment (benfluleck) [inline server/fixtures/user.js:24]: space needed here

File: server/fixtures/user.js [hunk 33]
```diff
@@ -1,9 +1,26 @@
...
 };
-export default newUser;
+export const goodUserData = {
+  email: 'testing@authorshaven.com',
+  username: 'testing',
+  lastName: 'testing',
+  firstName: 'testing',
+  password: 'testing'
+};
+export const emptyUserData = {
```
Comment (benfluleck) [inline server/fixtures/user.js:18]: space needed here

File: server/controllers/signup.js [hunk 34]
```diff
@@ -0,0 +1,30 @@
+import { cookieGenerator } from '<helpers>/utils';
+import model from '../models';
+
+const { User } = model;
+
+export const signUp = async (req, res) => {
+  const { formattedValues } = req;
+  const userData = {
```
Comment (benfluleck) [inline server/controllers/signup.js:13]: Please put a space here

File: server/fixtures/user.js [hunk 35]
```diff
@@ -1,9 +1,26 @@
...
+export const newUser = {
   email: 'lewislulu@yahoo.com',
   username: 'lulu',
   firstName: 'Lewis',
   lastName: 'Thomas',
   bio: 'Welcome to my world.',
   password: '123456'
 };
-export default newUser;
+export const goodUserData = {
```
Comment (benfluleck) [inline server/fixtures/user.js:10]: space needed here

**Comment by benfluleck** [comment]
small change and coverage has dropped

**Review by benfluleck — CHANGES_REQUESTED** [review]
I will check back tomorrow

File: server/helpers/utils.js [hunk 36]
```diff
@@ -1,17 +1,47 @@
...
  * @param {string} title
  * @return {string} unique_slug
  */
-const createUniqueSlug = title => `${slug(title, { lower: true })}-${Date.now()}`;
 
 /**
  * @param {string} password
  * @return {string} hash
  */
+const { SECRET, TOKENEXPIRYDATE, COOKIEEXPIRYDATE } = process.env;
```
Comment (benfluleck) [inline server/helpers/utils.js]: use underscore to space your constant names out `TOKEN_EXPIRY_DATE` and `COOKIE_EXPIRY_DATE`

File: server/helpers/utils.js [hunk 37]
```diff
@@ -1,17 +1,47 @@
...
+env.config();
 /**
  * @param {string} title
  * @return {string} unique_slug
  */
-const createUniqueSlug = title => `${slug(title, { lower: true })}-${Date.now()}`;
 
 /**
  * @param {string} password
  * @return {string} hash
```
Comment (benfluleck) [inline server/helpers/utils.js:7]: JSDoc is out of place

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/helpers/utils.js [hunk 38]
```diff
@@ -1,17 +1,58 @@
...
-const createUniqueSlug = title => `${slug(title, { lower: true })}-${Date.now()}`;
+export const createUniqueSlug = title => `${slug(title, { lower: true })}-${Date.now()}`;
 
 /**
- * @param {string} password
- * @return {string} hash
+ * @param {object} req
+ * @return {object} formattedObject
  */
+export const toLowerCaseAndTrim = (req) => {
```
Comment (benfluleck) [inline server/helpers/utils.js]: I have given you feedback this already, `req` cannot be an argument here. Change to something with a much more generic name.

File: server/tests/controller/user.spec.js [hunk 39]
```diff
@@ -0,0 +1,80 @@
...
+    chai
+      .request(app)
+      .post('/api/v1/auth/signup')
+      .send(emptyUserData)
+      .end((err, res) => {
+        expect(res.status).to.be.equal(400);
+        done();
+      });
+  });
+  it('should not allow invalid data types', (done) => {
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js:21]: Put a space here between it blocks

File: server/tests/controller/user.spec.js [hunk 40]
```diff
@@ -0,0 +1,80 @@
...
+    chai
+      .request(app)
+      .post('/api/v1/auth/signup')
+      .send(getUserData({ firstName: 666, lastName: 666, username: 99999 }))
+      .end((err, res) => {
+        expect(res.status).to.be.equal(400);
+        done();
+      });
+  });
+  it('should create a new user and save a token in the cookie', (done) => {
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js:32]: Put a space here between it blocks

File: server/tests/controller/user.spec.js [hunk 41]
```diff
@@ -0,0 +1,80 @@
...
+        expect(res.status).to.be.equal(201);
+        expect(res).to.be.an('object');
+        expect(res).to.have.cookie('access_token');
+        expect(res.body)
+          .to.have.property('message')
+          .equal('User created successfully');
+        done();
+      });
+  });
+  it('should not allow duplicate email and username', (done) => {
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js:48]: Put a space here between it blocks

File: server/tests/controller/user.spec.js [hunk 42]
```diff
@@ -0,0 +1,80 @@
...
+        expect(res.status).to.be.equal(409);
+        done();
+      });
+  });
+  it('should not allow duplicate email', (done) => {
+    chai
+      .request(app)
+      .post('/api/v1/auth/signup')
+      .send(getUserData({ email: 'testing@authorshaven.com', username: 'simi' }))
+      .end((err, res) => {
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js:64]: Put a space here between it blocks

File: server/tests/controller/user.spec.js [hunk 43]
```diff
@@ -0,0 +1,80 @@
...
+        expect(res.status).to.be.equal(409);
+        expect(res.body.message).to.equal('Email exist');
+        done();
+      });
+  });
+  it('should not allow duplicate username', (done) => {
```
Comment (benfluleck) [inline server/tests/controller/user.spec.js:71]: Put a space here between it blocks

**Review by benfluleck — CHANGES_REQUESTED** [review]

**Review by benfluleck — APPROVED** [review]
